Bfp16 gemm examples (#2436)
Xilinx/mlir-aie: programming_examples/ml/block_datatypes/matrix_multiplication/README.md

<img src="./res/gemm_dlt2_vis1.png" alt="Visualization of dlt2" width="400"/>
<img src="./res/gemm_dlt2_vis2.png" alt="Visualization of dlt2 bis" width="400"/>

The (impossible) equivalent transformation for block datatypes:

<img src="./res/gemm_dlt2_prbl1.png" alt="Visualization of dlt2 issue" width="400"/>
<img src="./res/gemm_dlt2_prbl2.png" alt="Visualization of dlt2 issue bis" width="400"/>
</details>

### Core Computations

Once the data has reached a compute tile, block datatypes also have additional complexities. They require additional manipulation in order to be loaded in and stored out of registers. The additional manipulations required to achieve this can be seen in [mm.cc](./mm.cc).

The main difference is due to the difficulty of loading the block_vector into the appropriate registers (illustrated below). To fully understand this problem, it is important to keep in mind how memory is accessed by the cores to load into the registers. For more information, please consult the [programming guide](../../../../programming_guide/section-4/section-4c/README.md).
<img src="./res/core_alignment.png" alt="Representation of difficulties of bfp vector loading"/>

## Examples

> Note that these examples are meant to be orientative and do not aim at being an ideal implementation. They may be used to evaluate the cost and capabilities of different operations and datatypes in the NPU, but should still be worked on to achieve maximal performance.

Reviewer: ```suggestion
> NOTE: These examples are meant to be orientative and do not aim at being an ideal implementation. They may be used to evaluate the cost and capabilities of different implementations and datatypes in the NPU, but could be developed further to achieve maximal performance.

```
